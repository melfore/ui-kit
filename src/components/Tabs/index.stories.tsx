import React from "react";
import { Grid } from "@material-ui/core";

import { getAllComposedDataCy } from "../../utils";
import { getDocumentationPage } from "../../utils/stories";
import Typography from "../Typography";

import Tabs, { DATA_CY_DEFAULT, SUBPARTS_MAP } from ".";

export default {
  title: "Tabs",
  component: Tabs,
  parameters: {
    ...getDocumentationPage({
      basedOn: "@material-ui/core/Tabs",
      component: "Tabs",
      e2eTestInfo: {
        dataCyDefault: DATA_CY_DEFAULT,
        subpartsSuffixes: getAllComposedDataCy(SUBPARTS_MAP),
      },
    }),
  },
};

const tabs = [
  {
    component: (
      <div style={{ padding: "16px" }}>
        <Grid container spacing={4}>
          <Grid item xs={12}>
            <Typography>Mosaic Library</Typography>
          </Grid>
        </Grid>
      </div>
    ),
    label: "Mosaic",
  },
  {
    component: (
      <div style={{ padding: "16px" }}>
        <Grid container spacing={4}>
          <Grid item xs={12}>
            <Typography>Components List</Typography>
          </Grid>
        </Grid>
      </div>
    ),
    label: "Components",
  },
];

export const Canvas = () => <Tabs tabs={tabs} />;

export const Basic = () => <Tabs tabs={tabs} />;

export const DefaultTab = () => <Tabs defaultTab={1} tabs={tabs} />;
