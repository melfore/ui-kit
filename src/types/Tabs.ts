import { IBase } from "./Base";

interface ITab {
  component: JSX.Element;
  label: string;
}

export interface ITabs extends IBase {
  defaultTab?: number;
  tabs: ITab[];
}
