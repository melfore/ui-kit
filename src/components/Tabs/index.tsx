import React, { FC, useState } from "react";
import { Grid as MUIGrid, Paper as MUIPaper, Tab as MUITab, Tabs as MUITabs } from "@material-ui/core";

import { ITabs } from "../../types/Tabs";
import { getComposedDataCy } from "../../utils";

export const DATA_CY_DEFAULT = "tabs";

export const SUBPARTS_MAP = {
  tab: {
    label: "Tab",
    value: (label = "{label}") => `tab-${label}`,
  },
  tabContent: {
    label: "Tab Content",
  },
};

// TODO#lb: handle style
const Tabs: FC<ITabs> = ({ dataCy = DATA_CY_DEFAULT, defaultTab = 0, tabs }) => {
  const [tab, setTab] = useState(defaultTab);

  return (
    <MUIGrid item xs={12}>
      <MUIPaper>
        <MUITabs
          data-cy={dataCy}
          indicatorColor="primary"
          onChange={(_, tab: number) => setTab(tab)}
          textColor="primary"
          value={tab}
        >
          {tabs.map(({ label }) => (
            <MUITab data-cy={getComposedDataCy(dataCy, SUBPARTS_MAP.tab, label)} key={label} label={label} />
          ))}
        </MUITabs>
        <MUIGrid data-cy={getComposedDataCy(dataCy, SUBPARTS_MAP.tabContent)} item xs={12}>
          {tabs[tab].component}
        </MUIGrid>
      </MUIPaper>
    </MUIGrid>
  );
};

export const TabsProps = Tabs;

export default Tabs;
