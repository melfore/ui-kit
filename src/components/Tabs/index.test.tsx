import React from "react";
import renderer from "react-test-renderer";

import { ITabs } from "../../types/Tabs";
import { getTestable } from "../../utils/tests";

import Tabs, { DATA_CY_DEFAULT } from ".";

const defaultProps: ITabs = {
  tabs: [
    {
      component: <div>Tab 0</div>,
      label: "Tab0",
    },
    {
      component: <div>Tab 1</div>,
      label: "Tab1",
    },
  ],
};

const getTabsTestable = (props?: ITabs, dataCy = DATA_CY_DEFAULT) =>
  getTestable(Tabs, { dataCy, domNode: "div", props: { ...defaultProps, ...props } });

describe("Tabs test suite:", () => {
  it("default", () => {
    const { element, wrapper } = getTabsTestable();
    expect(wrapper).toHaveLength(1);
    expect("default-props-check").toBeTruthy();

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("dataCy", () => {
    const { element, wrapper } = getTabsTestable({ ...defaultProps, dataCy: "custom" }, "custom");
    expect(wrapper).toHaveLength(1);

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("tab change", () => {
    const { element, wrapper } = getTabsTestable();
    const tab1Button = wrapper.find(`button[data-cy='${DATA_CY_DEFAULT}-tab-Tab1']`);
    tab1Button.simulate("click");

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  it("tab default", () => {
    const { element, wrapper } = getTabsTestable({ ...defaultProps, defaultTab: 1 });
    const tab1Button = wrapper.find(`button[data-cy='${DATA_CY_DEFAULT}-tab-Tab1']`);
    expect(tab1Button.hasClass("Mui-selected"));

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });
});
