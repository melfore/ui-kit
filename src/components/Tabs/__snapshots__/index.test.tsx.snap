// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs test suite: dataCy 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiTabs-root"
      data-cy="custom"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": null,
            "overflow": "hidden",
          }
        }
      >
        <div
          className="MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
            data-cy="custom-tab-Tab0"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Tab0
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            data-cy="custom-tab-Tab1"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Tab1
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          className="PrivateTabIndicator-root-1 PrivateTabIndicator-colorPrimary-2 MuiTabs-indicator"
          style={
            Object {
              "left": 0,
              "width": 0,
            }
          }
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      data-cy="custom-tab-content"
    >
      <div>
        Tab 0
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs test suite: default 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiTabs-root"
      data-cy="tabs"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": null,
            "overflow": "hidden",
          }
        }
      >
        <div
          className="MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
            data-cy="tabs-tab-Tab0"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Tab0
            </span>
            <span
              className="PrivateTabIndicator-root-1 PrivateTabIndicator-colorPrimary-2 MuiTabs-indicator"
              style={Object {}}
            />
          </button>
          <button
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            data-cy="tabs-tab-Tab1"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Tab1
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      data-cy="tabs-tab-content"
    >
      <div>
        Tab 0
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs test suite: tab change 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiTabs-root"
      data-cy="tabs"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": null,
            "overflow": "hidden",
          }
        }
      >
        <div
          className="MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
            data-cy="tabs-tab-Tab0"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Tab0
            </span>
            <span
              className="PrivateTabIndicator-root-1 PrivateTabIndicator-colorPrimary-2 MuiTabs-indicator"
              style={Object {}}
            />
          </button>
          <button
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            data-cy="tabs-tab-Tab1"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Tab1
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      data-cy="tabs-tab-content"
    >
      <div>
        Tab 0
      </div>
    </div>
  </div>
</div>
`;

exports[`Tabs test suite: tab default 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiTabs-root"
      data-cy="tabs"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": null,
            "overflow": "hidden",
          }
        }
      >
        <div
          className="MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            data-cy="tabs-tab-Tab0"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Tab0
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
            data-cy="tabs-tab-Tab1"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Tab1
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          className="PrivateTabIndicator-root-1 PrivateTabIndicator-colorPrimary-2 MuiTabs-indicator"
          style={
            Object {
              "left": 0,
              "width": 0,
            }
          }
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      data-cy="tabs-tab-content"
    >
      <div>
        Tab 1
      </div>
    </div>
  </div>
</div>
`;
